### ISSUE TITLE
[Enhancement proposal] Exception interceptors

### ISSUE THREAD

**Issue by adamecr** [safakgur/guard#21]
Hello @safakgur , first of all, thanks for a great library, I like the design and have tried to use it in some of my projects.
I have played a bit and found quite tricky to log the exception before it's actually thrown. Actually, I was able to do it using logger extension
```csharp
public delegate T Dlgt<T>( Guard.ArgumentInfo<T> argument);
public static T Guard<T>(this ILogger logger, T a, string name, Dlgt<T> f)
{
    try
    {
        return f.Invoke(Dawn.Guard.Argument(a, nameof(a)));
    }
    catch (Exception ex)
    {
        logger.Fatal(exception: ex, ex.Message);
        throw;
    }
}
```
and then
```csharp
string bb = Logger.Guard(() => Guard.Argument(a, nameof(a)).NotWhiteSpace().NotNull());
```

But it makes the whole argument validation less readable and prone to human errors.

My proposal is to extend the `ArgumentInfo` with the ability to keep the reference to exception interceptor and call such interceptor before the exception is thrown when defined.
The validation will look like this (of course it's even possible to have more overloads to simplify it a bit more): 
```csharp
string bb = Guard.Argument(a, nameof(a), exceptionInterceptor:(ex) =>logger.Fatal(ex)).NotWhiteSpace().NotNull();
```
I have tried to implement it [at my fork](https://github.com/adamecr/guard/commit/75a9132b1e595d2060c1cdc4d7fa2e31e09856ab), and created the [pull request](https://github.com/safakgur/guard/pull/22). 

The main changes are in `AgrumentInfo<T>` class. Plus added new parameters to`Guard.Argument` and ensured that the `ArgumentInfo` is copied with `exceptionInterceptor` when modified. The rest is just about wrapping the exception into `throw argument.Exception(...)`.

**Comment by safakgur** [comment]
Hi @adamecr - I'm glad you liked the library, thank you for using it.

Intercepting the validation exceptions sounds like an interesting concept, but I don't think specifying the interceptors per-argument provides a clean API surface. Here are some of my thoughts about the proposal:

* I expect all validations in a single method to use the same exception interceptor in most cases. With this implementation, however, the usage of exception interceptors is duplicated for each argument we validate. So, if we're validating more than one argument, just wrapping all the validations in a single try-catch block would be more readable.
* The interceptor delegate would likely capture outside state (like the logger instance in your sample), causing a new allocation with each validation. Using a try-catch instead would prevent this, too.

```c#
Guard.Argument(a, nameof(a), exceptionInterceptor: ex => logger.Fatal(ex)).NotNull().NotWhiteSpace();
Guard.Argument(b, nameof(b), exceptionInterceptor: ex => logger.Fatal(ex)).NotNull().NotWhiteSpace();

// vs.

Action<Exception> interceptor = ex => logger.Fatal(ex);
Guard.Argument(a, nameof(a), exceptionInterceptor: interceptor).NotNull().NotWhiteSpace();
Guard.Argument(b, nameof(b), exceptionInterceptor: interceptor).NotNull().NotWhiteSpace();

// vs.

try
{
    Guard.Argument(a, nameof(a)).NotNull().NotWhiteSpace();
    Guard.Argument(b, nameof(b)).NotNull().NotWhiteSpace();
}
catch (ArgumentException ex)
{
    logger.Fatal(ex);
}
```

Regarding the logging example, I'm not sure if the exact place where we perform argument validations is right for logging exceptions. I predict this to be a task that is more suitable for the calling site, or even higher in the stack. Think Int32.Parse for an exaggerated example; it does not know anything about any loggers, it just throws the relevant argument exceptions for invalid data and the caller of this method is responsible for logging the result or the exception.

If we are to introduce exception interceptors into Guard, I believe it should be a little more convenient to use. What would you think of having scoped exception interceptors? We can use the logical call context under the hood to implement something like the following, I suppose.

```c#
// You can create an application-wide scope in Main or
// a request-wide scope in BeginRequest like this:
Guard.BeginExceptionInterceptorScope(ex => { /* ... */ });

// Or you can create an explicit scope like this:
using (Guard.BeginExceptionInterceptorScope(ex => { /* ... */ }))
{
    // Guard statements.
}
```

Something like this would allow us to create application-wide, request-wide (for web services) or explicitly scoped exception interceptors. I don't see people providing different exception interceptors per-argument anyway. A more likely example I can think of is creating request-wide scopes that log the exceptions along with unique request IDs.

Thank you for the idea, I'm looking forward to hearing your thoughts on this.

P.S. Jon Skeet has a [great post][1] about implicit context where he lists some of its pros and cons in detail. It's a nice read.

[1]: https://codeblog.jonskeet.uk/2010/11/08/the-importance-of-context-and-a-question-of-explicitness/

**Comment by adamecr** [comment]
Hello @safakgur , thanks a lot for the feedback. It's always good to have the possibility to discuss the ideas.

I agree that providing the interceptor per-validation is not quite effective and like the idea using the delegate (I don't use try/catch as I don't really want to catch the exception after logging and re-throwing messes up the stack a bit).
Just to explain a bit more the use case with logging - I use it for the diagnostic logging of the components, so even if the exception is handled on some higher level, I'd like to know that there might be some misuse of the component or the use case I didn't think about causing the "unexpected arguments". Well, even for this case, you are right, it doesn't make sense to provide the interceptors per-argument.

I must admit, I like the idea about using the scoped approach!
I have played a bit and come with following:
- Each `ArgumentInfo` belongs to single scope
- There is a root (default) scope managing the argument infos created by `Guard.Argument`.
  - The root scope has no exception interceptor
- Each scope can have multiple child scopes and the scopes can be nested
  - The child scope can have the exception interceptor (but is not mandatory)
  - The child scope can create argument infos using `scope.Argument<T>` method. Such argument infos are assigned to this child scope
- Each scope is implemented as disposable, when the scope is disposed, the children tree is disposed as well
  - Touching the exception interceptor managed by disposed scope throws `ObjectDisposedException`

Note: I have tried not to bind the scopes directly to the exception interceptors, so the scope can be (maybe) used also for something else (for example some kind of aggregating the exceptions came to my mind).

A few examples:
Create the argument info belonging to the root scope (no interceptors available):
```
Guard.Argument(() => argument)
```

Create the child scope of the root, assign it with the interceptor and create the argument info belonging to the child scope (the exceptions will be intercepted):
```
using (var scope = Guard.BeginScope(ex => ... ))
{
    var emptyArg = scope.Argument(() => argument);
 }
```

Create the nested child scope with the interceptor and create the argument info belonging to the nested scope (the exceptions will be intercepted by "inner" interceptor):
```
using (var scope = Guard.BeginScope(ex => ... ))
{
    using (var scopeInner= scope .BeginScope(ex => ... ))
    {
        var emptyArg = scopeInner.Argument(() => argument);
    }
 }
```
More examples can be seen in the [tests](https://github.com/adamecr/guard/blob/feature/scoped-interceptors/tests/InterceptorBasicTests.cs).

Something I'd like to discuss whether it's a good idea or should be adjusted:
- The `ArgumentInfo.ExceptionInterceptor` is currently public. It works also when it's private, so maybe private/protected modifier will be better. I'm just not sure whether limiting the access will not also limit the potential extensibility (vs. potential misuse/confusion when public).
- The exception is intercepted only within the scope the invalid argument belongs to and is not propagated to parent scope interceptors.
- When there is no interceptor in the scope of the invalid argument, it doesn't traverse the scope tree up to check whether there is any parent interceptors (no "bubble up").
Note: this is the reason I don't publish `GuardScope.Parent` to `IGuardScope` now.

The implementation is available in the [feature branch](https://github.com/adamecr/guard/tree/feature/scoped-interceptors):
- `IGuardScope` is the public interface of the scope, allowing to
   - Create the child scope (`BeginScope` method)
   - Create scope bound argument info (`Argument<T>` methods)
   - Get the exception interceptor (`ExceptionInterceptor` property)
- `GuardScope` is the implementation of `IGuardScope`
  - Manages the root scope (kind of singleton)
  - Manages (and disposes) the child scopes
- `GuardDisposable` and `GuardDisposer` are helpers for disposing the objects and scopes
- `Guard.Scope` (partial Guard) provides the `BeginScope` method allowing to create the child scope of the root scope
- `ArgumentInfo` contains `Scope` property
  - `ArgumentInfo.ExceptionInterceptor` is mapped to `ArgumentInfo.Scope.ExceptionInterceptor`
- `InterceptorBasicTests` are implemented for the use cases that I have thought about during the design.

That's it for the moment, l'm curious whether it fits to your idea.
I'm closing the original pull request as being very premature and let's see where we will get after the discussion.

P.S. Just thinking that the exception interceptors could be used not just for logging but may provide some corrections (of course this will mean changes in signature), on the other hand, this kind of implicit changes is quite dangerous and is probably an anti-pattern. But I can imagine that it might be safe/useful in some cases like "if less than `min`, set to `min`" or "if empty, set to some default". Anyway, just an idea for future :smile:

**Comment by safakgur** [comment]
Nice work, though I believe that the validating code shouldn't need to know anything about scopes:

```c#
void Foo(string s)
{
    try
    {
        // The scopes are transparent for the validating code.
        Guard.Argument(() => s).NotNull();
    }
    catch (ArgumentNullException)
    {
    }
}

void Bar()
{
    // This should print nothing since we haven't started a scope yet.
    Foo(null);

    // Start a scope.
    Guard.BeginScope(ex => Console.WriteLine("A"));

    // This should print "A".
    Foo(null);

    // Start an explicit scope.
    using (Guard.BeginScope(ex => Console.WriteLine("B")))
    {
        // This should print "B".
        Foo(null);
    }

    // At the end of the using block, the explicit scope is disposed.
    // So this should print "A" again.
    Foo(null);
}
```

We don't need argument instances to belong to scopes, we only need to check if there is a "current" scope exists at the time of validation failure. I [took a stab][1] at implementing this using [`AsyncLocal<T>`][2].

* > The ArgumentInfo.ExceptionInterceptor is currently public.

  It feels wrong to expose the exception interceptor as a property of `ArgumentInfo<T>` since it belongs to the scope and not the argument itself.

* > The exception is intercepted only within the scope the invalid argument belongs to and is not propagated to parent scope interceptors.

  I've also implemented the PoC this way for the sake of simplicity but I assume most people would expect a scope to work everywhere it encloses. I think we should make propagation the default behavior and consider adding an optional, boolean parameter to `BeginScope`, so the users can opt-out.

* > When there is no interceptor in the scope of the invalid argument, it doesn't traverse the scope tree up to check whether there is any parent interceptors (no "bubble up").

  I don't think scope having an exception interceptor or not should affect whether we should traverse the tree.

* > IGuardScope is the public interface of the scope, allowing to
  > * Create the child scope (BeginScope method)
  > * Create scope bound argument info (Argument<T> methods)

  I'm in favor of making the scopes transparent, so creating child scopes or validating arguments in a scope does not require the scope instance itself.

  > * Get the exception interceptor (ExceptionInterceptor property)
  
  I don't see value in exposing the exception interceptor. And with all the other members gone, there is no reason left to expose a scope object at all. `BeginScope` can just return an `IDisposable` that ends the scope when disposed. It would be an easy and non-breaking change to return a more-specific implementation in the future.

* > Just thinking that the exception interceptors could be used not just for logging but may provide some corrections

  This sounds very interesting - but  I'm afraid changing the behavior of the validations by some implicit, invisible scope would cause more confusions than the problems it solves :). We can explore more uses with the scoping infrastructure in place.

What do you think about having transparent scopes? Do you find the ones that explicitly own arguments like in your branch more clear?
 
[1]: https://github.com/safakgur/guard/pull/23/files
[2]: https://docs.microsoft.com/dotnet/api/system.threading.asynclocal-1

**Comment by adamecr** [comment]
Frankly, I think I have never implemented transparent scopes in my code - mostly because I like having control about "what happens where" and I use multiple sibling sub-scopes quite often (so sub-scopes are more like a tree than chain).

However, when re-thinking the idea, I have found a show-stopper for the scope-owned arguments - the guards (validations) in existing code will always be out of scopes and the adjustment will mean quite a rework of existing components that are using Guard. It will be backwards compatible, but adopting the functionality will be too costly for the purpose, I guess.

So finally, I agree that transparent scopes are the neat way of implementation for Guard. Also the way you have used in the PoC is fine and nice - chain of sub-scopes based on `AsyncLocal<T>` is good enough for the case, simple to implement and understand, and eliminates the need to touch the ArgumentInfo at all.

I used your PoC and extended it to other validations to be able to check it with some of the code I use the forked Guard for. I have [adjusted the code](https://github.com/adamecr/guard/tree/check/scopes) just for one case - sometimes I need to "disable" the interceptors in some of the inner scopes. I didn't want to do it using the interceptor that "does nothing", so I added the optional parameter to BeginScope allowing to use "forced null interceptor" (the scope is created, but the interceptor is null). 

With that simple adjustment, I have been able to fulfill all the needs, so it's a definitely go for me.

While extending the `throw Fail(...)` to other places, I added the interception to the argument validations only and left the "internal" Guard exceptions not intercepted. For example in `Guard.Member`, I do intercept member validation exceptions, but don't intercept the validations of own Guard.Member arguments. It feels a bit better to me, but didn't make up my mind yet :)

P.S. I think that the `id` should be reset before each `Test(...)` in the `ScopeTests` to ensure the proper beginning state - [updated here](https://github.com/adamecr/guard/blob/check/scopes/tests/ScopeTests.cs).

**Comment by safakgur** [comment]
Thank you, I think we're almost there.

1. I agree that the internal guard exceptions should not be intercepted. 👍
2. Since we intercepted the exceptions before they are thrown, they contained no stack trace at the interception, making them somewhat useless. I started throwing in Fail (d215dd4b509b7636e3ec046545a5c04abaa033db) to get the stack trace.
3. I made the exception interceptors bubble up by default (3991826a49f7f1b48a48baa20a0de8675cbfcdc9), as we discussed earlier.
4. Your forced null interceptor allowed us to disable bubbling, but I thought we should cover all the following:
   * Inner scope with an interceptor that bubbles up to outer scopes.
   * Inner scope with an interceptor that does not bubble up to outer scopes.
   * Inner scope without an interceptor that bubbles up to outer scopes.
   * Inner scope without an interceptor that does not bubble up to outer scopes.
   
   &nbsp;
   So I changed it to a boolean parameter - 5c39d955831292855adb1f28481b4b3e1e0fb28f

Do you think there is anything left we didn't cover?

**Comment by adamecr** [comment]
Looks good, I think we have covered even more than I have had in my mind at the beginning 👍 

just ad point 2:
Actually, having no stack trace in my PoC was one of the weak points that I was not able to solve a good way, so I intentionally left the stack trace empty for the interceptors. I tried to use the similar solution (throw in Fail), however, when you check in the interceptor, the stack trace contains just single line about the Fail method, but not the full stack.

It passes the test, because the interceptor in the test just keeps the reference to the event instance and when Assert checks the exception it is already manipulated after the interceptor (full stack added on the [re]throw). So the reference is the same, but the content is not the same as has been in time of the interception (you can simply check by storing also the stack trace in the interceptor).

The only so-so working solution was to
- throw the exception in the validator
- catch it there
- capture the exception using ExceptionDispatchInfo.Capture
- send to interceptors
- re-throw  using ExceptionDispatchInfo.Throw

The so-so means that the stack trace was at least about the validator, however I think that the stack trace is built when the exception bubble backwards until it's catch. So we will probably not be able to get the full stack trace (call history) on this level, I'm afraid.

As this felt quite complicated, so I gave up the stack trace for the PoC...
... and forget about it until now

I'll try to think about a bit more, but any ideas are welcome
[EDIT] Actually, got some idea, checking...

**Comment by safakgur** [comment]
Excellent finding. I think we can also consider changing the interceptor type to `Action<Exception, string>`, so we can send [Environment.StackTrace][1] along with the exception.

[1]: https://docs.microsoft.com/dotnet/api/system.environment.stacktrace

**Comment by adamecr** [comment]
Actually, I have had similar idea, but you nailed it!
I was not aware about the `Environment.StackTrace` at all, so I was just checking the PoC of building the stack using `new StackTrace(0).GetFrames()` 😄

**Comment by adamecr** [comment]
Just to add something to brainstorming - two more promising ideas I have had in past:
- Create own Exception type with stack trace in ctor and override the StackTrace property - I didn't like it, because the argument validation exceptions are well know and this will bring a "fake" ones
- Add the stack trace to exception instance using the reflection - I have made it working even the Visual Studio debugger accepted it after throwing, but I think this is very dangerous misuse of the reflection.

So adding the stack trace as additional parameter to interceptor still sounds good to me

**Comment by safakgur** [comment]
Regarding the first idea, we could pass a custom exception to the interceptor with its inner exception having the original exception that is initialized by the validator. The exception we actually throw wouldn't change. But this would confuse the people creating scopes.

You are right about the second idea being dangerous, we shouldn't depend on implementation details.

So I think sending the stack trace as a separate argument is the most explicit/clear approach we can take. It also means we no longer need to throw on Fail.

I just committed this to the scopes branch. Since we're no longer throwing on Fail, your updates to the rest of the validators should work. You can open a PR into the scopes branch with those changes and I'd be happy to accept it.

Edit: I just remembered that I rebased the scopes branch from dev. I'm sorry, I'm so used to working on this alone, it didn't occurred to me how this would affect your changes. 🤦‍♂️

**Comment by adamecr** [comment]
PR created. No problem with rebased branches, it was simple to apply, but thanks for the notice.

These are the exceptions not being intercepted:
- `Guard.Enums.Legacy.cs` - it's obsolete
- `Guard.Member.cs` - `public static ref readonly ArgumentInfo<T?> Member<T, TMember>(...)` - doesn't intercept exceptions in _Get member info_ and _Get member value_ parts (intercept in _Validate the member_)
- `Guard.Member.cs` - doesn't intercept exceptions in `private abstract class ArgumentMemberInfo`
- `Guard.Require.cs` - doesn't intercept exceptions in `private static class Exception<T>`
- `Guard.String.cs` - doesn't intercept regex exceptions in `Matches` and `DoesNotMatch` (intercepted in `if(!matches)` resp. `if(matches)`)

I hope I didn't miss something.

**Comment by safakgur** [comment]
It looks perfect - thank you for all the work you've put in!

I'll make Fail public to support extensions and update the extensibility document accordingly. After that, I believe we're ready for 1.7.0.

**Comment by adamecr** [comment]
Superb, thanks a lot for the support with the request. Not only it will help me using Guard, I have also enjoyed the design discussion and learnt something new.

Looking forward 1.7.0 👍 

P.S. If you don't mind, I'll keep the ticket open until 1.7.0 release not to confuse the other visitors

**Comment by adamecr** [comment]
Hello @safakgur ,
just played a bit with the stack trace and I might have one more proposal - using `new StackTrace(1, true).ToString()` instead of  `Environment.StackTrace` while invoking the interceptor in `Guard.Fail` can point to the "real" origin (validation) as it skips the stack trace frames of stack trace getter and the Fail method itself.
If it make sense to you, it's ready for PR at [de2ea61](https://github.com/adamecr/guard/commit/de2ea61598c50931a09be2ab717c43fb6a3eaa91), but it's nothing critical, just a small adjustment.

`Environment.StackTrace` returns:
```
   at System.Environment.get_StackTrace()
   at Dawn.Guard.Fail(Exception exception) in C:\Users\radek\Documents\Visual Studio 2017\Projects\fork.safakgur.guard\guard\src\Guard.Argument.cs:line 75
   at Dawn.Guard.NotNull[T](ArgumentInfo`1& argument, String message) in C:\Users\radek\Documents\Visual Studio 2017\Projects\fork.safakgur.guard\guard\src\Guard.Null.cs:line 95
   at Dawn.Tests.ScopeTests.<BeginScope>g__Test|0_1(Exception& intercepted) in C:\Users\radek\Documents\Visual Studio 2017\Projects\fork.safakgur.guard\guard\tests\ScopeTests.cs:line 124
   at Dawn.Tests.ScopeTests.<>c.<<BeginScope>b__0_3>d.MoveNext() in C:\Users\radek\Documents\Visual Studio 2017\Projects\fork.safakgur.guard\guard\tests\ScopeTests.cs:line 34
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder.Start[TStateMachine](TStateMachine& stateMachine)
   at Dawn.Tests.ScopeTests.<>c.<BeginScope>b__0_3()
   at System.Threading.Tasks.Task`1.InnerInvoke()
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Threading.Tasks.Task.ExecuteWithThreadLocal(Task& currentTaskSlot)
   at System.Threading.ThreadPoolWorkQueue.Dispatch()
```
`new StackTrace(1, true).ToString()` returns:
```
   at Dawn.Guard.NotNull[T](ArgumentInfo`1& argument, String message) in C:\Users\radek\Documents\Visual Studio 2017\Projects\fork.safakgur.guard\guard\src\Guard.Null.cs:line 95
   at Dawn.Tests.ScopeTests.<BeginScope>g__Test|0_1(Exception& intercepted) in C:\Users\radek\Documents\Visual Studio 2017\Projects\fork.safakgur.guard\guard\tests\ScopeTests.cs:line 124
   at Dawn.Tests.ScopeTests.<>c.<<BeginScope>b__0_3>d.MoveNext() in C:\Users\radek\Documents\Visual Studio 2017\Projects\fork.safakgur.guard\guard\tests\ScopeTests.cs:line 34
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder.Start[TStateMachine](TStateMachine& stateMachine)
   at Dawn.Tests.ScopeTests.<>c.<BeginScope>b__0_3()
   at System.Threading.Tasks.Task`1.InnerInvoke()
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Threading.Tasks.Task.ExecuteWithThreadLocal(Task& currentTaskSlot)
   at System.Threading.ThreadPoolWorkQueue.Dispatch()
```

**Comment by safakgur** [comment]
Looks great, go ahead.

**Comment by adamecr** [comment]
Thanks for merging. I don't have anything more in mind. Also checked the other commits and I think that the feature is well prepared. Good job 👍 , thank you.

**Comment by safakgur** [comment]
Thank you for the PR.

Since we're already creating a `StackTrace` instance, I'm thinking maybe we should pass it directly to the interceptor. The user can call `ToString` if they want. This would would also allow them to enumerate frames to programmatically request details like row/col numbers or `MethodInfo`s.

What do you think?

**Comment by adamecr** [comment]
That's a good point. It will bring additional flexibility to the interceptors without big effort.

**Comment by adamecr** [comment]
Implemented in version [1.7.0](https://github.com/safakgur/guard/tree/v1.7.0), closing the ticket.
Thanks a lot @safakgur
